Taylor and Francis Group.js

```diff
@@ -0,0 +1,287 @@
+		});
+	}
+	else {
+		scrape(doc, url, itemType);
+	}
+}
+
+function scrape(doc, url, itemType) {
+	if (!itemType) {
+		itemType = detectWeb(doc, url);
+	}
+
+	let item = new Zotero.Item(itemType);
+	item.title = text(doc, 'h1.product-banner-title');
+	item.libraryCatalog = '';
+	item.extra = '';
+	let subtitle = text(doc, '.product-banner-subtitle');
+	if (subtitle) {
+		item.title += ": " + subtitle;
+	}
+	if (itemType == 'bookSection') {
+		item.bookTitle = text(doc, 'a.product-book-link');
+	}
+	
+	for (let creator of text(doc, '.product-banner-author-name').split(", ")) {
+		let creatorType = text(doc, '.product-banner-author').includes('Edited')
+			? 'editor'
+			: 'author';
+		item.creators.push(ZU.cleanAuthor(creator, creatorType, false));
+	}
+	
+	for (let row of doc.querySelectorAll('product-more-details .display-row')) {
+		let label = text(row, 'span:first-child');
+		let value = text(row, 'span:last-child');
+
+		let field = FIELDS[label];
+		if (field) {
+			if (field == 'extra') {
+				if (label == "DOI") {
+					value = value.match(/(?:doi\.org\/)?([^/]+\/[^/]+)/)[1];
+				}
+				item.extra += `${label}: ${value}\n`;
+			}
+			else {
+				item[field] = value;
+			}
+		}
+	}
+	
+	if (doc.cookie) {
+		let apiId = doc.getElementById('pyramid-website-ssr-state').textContent
+			.match(/&q;_id&q;:&q;([^&]+)&q;/)[1];
+		let token = doc.cookie
+			.split('; ')
+			.find(row => row.startsWith('_token='))
+			.split('=')[1];
+		ZU.doGet(`https://api.taylorfrancis.com/v4/content/${apiId}?apiVersion=4.0.1&filenamePrefix=${item.ISBN}&render=false`,
+			function (respText, xhr) {
+				// the request might fail if we don't have PDF access
+				// but that's alright, everything else worked
+				// so we treat every status as a success (last arg to doGet)
+				// but only try to fetch the attachment if we got 200 OK
+				
+				if (xhr.status == 200) {
+					let resp = JSON.parse(respText);
+					for (let contentObject of resp) {
+						if (contentObject.type == 'webpdf'
+							|| contentObject.type == 'chapterpdf') {
+							item.attachments.push({
+								title: "Full Text PDF",
+								mimeType: "application/pdf",
+								url: contentObject.location
+							});
+							break;
+						}
+					}
+				}
+				
+				item.complete();
+			}, null, null, {
+				Authorization: 'idtoken ' + token,
+				Referrer: 'https://www.taylorfrancis.com/'
+			}, false);
+	}
+	else {
+		item.complete();
+	}
+}
+
+/** BEGIN TEST CASES **/
+var testCases = [
+	{
+		"type": "web",
+		"url": "https://www.taylorfrancis.com/books/edit/10.1201/9781315120621/lubricant-additives-leslie-rudnick",
+		"items": [
+			{
+				"itemType": "book",
+				"title": "Lubricant Additives: Chemistry and Applications",
+				"creators": [
+					{
+						"firstName": "Leslie R.",
+						"lastName": "Rudnick",
+						"creatorType": "editor"
+					}
+				],
+				"date": "29 June 2017",
+				"ISBN": "9781315120621",
+				"edition": "3rd Edition",
```

Comment:
Check me on this, but I think the edition needs to be just a number to be properly handled by the citation processor, and otherwise it's passed through as a literal (meaning it won't be localized). If that's the case, when we can parse it cleanly here, we'd want to add it as just an integer.

Reply:
Oh, good to know. Will add some code to try to parse it (`cleanEdition` could be a good `Zotero.Utilities` function...).

Reply:
>  I think the edition needs to be just a number to be properly handled by the citation processor, and otherwise it's passed through as a literal

Basically, yes. The tricky part in parsing this is to not remove useful information: e.g. "2nd edition with a forword by the author" should be left as literal, so you don't just want to match on a number.
You also want to typically remove 1st editions (those are often in library catalogs) entirely.

Reply:
Oh, not removing useful information is a good point. I *think* I have it right as of the latest commit: I trim the edition text, discard it if it looks like a first edition, and then try to pull out just the number as long as it's followed only by a suffix and `ed.`/`edition`.

Now that I think about it, maybe I want to reverse the order so that we keep stuff like "1st edition with a foreword by the author." Not a common edge case, but it's possible.

Reply:
Yeah, a `cleanEdition()` seems like a good idea. If you write it as a separate function, we could bundle it and start including it as a shim in other translators.